pyro/poutine/trace.py

```diff
@@ -81,8 +82,10 @@ def log_pdf(self, site_filter=lambda name, site: True):
                     site_log_p = site["log_pdf"]
                 except KeyError:
                     args, kwargs = site["args"], site["kwargs"]
-                    site_log_p = site["fn"].log_prob(
-                        site["value"], *args, **kwargs).sum() * site["scale"]
+                    if isinstance(site["fn"], RandomPrimitive):
+                        site_log_p = site["fn"].log_prob(site["value"], *args, **kwargs).sum() * site["scale"]
+                    else:
+                        site_log_p = site["fn"].log_prob(site["value"]).sum() * site["scale"]
```

Comment:
`TorchDistribution` is the special case here.  This should be:
```
if isinstance(site["fn"], TorchDistribution):
    site_log_p = site["fn"].log_prob(site["value"]).sum() * site["scale"]
else:
    site_log_p = site["fn"].log_prob(site["value"], *args, **kwargs).sum() * site["scale"]
```

Reply:
Ok, Done.

I was hoping to remove this kludgey `isinstance()` logic when we remove `RandomPrimitive`. I think this would be possible if you refactored `Histogram` to be a distribution factory rather than an `Distribution`. That way our `Distribution` interface could be narrower and closer to the PyTorch `Distribution` interface, and `Histogram` would only differ from it's current behavior by a `functools.partial`. This would make Neeraj's and my lives easier. What do you think?

Reply:
I'd have to see the proposed changes to `Histogram` but it's probably fine

Reply:
See also #23 for a much older discussion about this